File: backend/app/schemas/recipe.py
"""Recipe request/response schemas."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class RecipeIngredientCreate(BaseModel):
    ingredient_id: int
    quantity: float = Field(gt=0)
    unit: str


class RecipeIngredientResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    ingredient_id: int
    ingredient_name: str
    quantity: float
    unit: str
    unit_cost: float | None
    unit_cost_unit: str | None = None
    line_cost: float | None


class RecipeCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    selling_price: float = Field(gt=0)
    category: str | None = None
    target_margin: float | None = None
    is_homemade: bool = True
    ingredients: list[RecipeIngredientCreate] = Field(default_factory=list)


class RecipeUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    selling_price: float | None = Field(default=None, gt=0)
    category: str | None = None
    target_margin: float | None = None
    is_homemade: bool | None = None
    ingredients: list[RecipeIngredientCreate] | None = None


class RecipeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    selling_price: float
    category: str | None
    target_margin: float | None
    food_cost: float | None
    food_cost_percent: float | None
    is_homemade: bool
    margin_status: str
    ingredients: list[RecipeIngredientResponse] = []
    created_at: datetime


class RecipeListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    selling_price: float
    category: str | None
    target_margin: float | None
    food_cost: float | None
    food_cost_percent: float | None
    is_homemade: bool
    margin_status: str
    created_at: datetime


class RecipeListResponse(BaseModel):
    items: list[RecipeListItem]
    total: int


class DashboardResponse(BaseModel):
    average_food_cost_percent: float | None
    total_recipes: int
    recipes_green: int
    recipes_orange: int
    recipes_red: int
    recipes: list[RecipeListItem]

File: frontend/src/types/recipe.ts
// Mirror of backend/app/schemas/recipe.py

export type MarginStatus = 'green' | 'orange' | 'red';

export interface RecipeIngredientResponse {
  id: number;
  ingredient_id: number;
  ingredient_name: string;
  quantity: number;
  unit: string;
  unit_cost: number | null;
  unit_cost_unit: string | null;
  line_cost: number | null;
  converted_quantity?: number | null;
  conversion_ok?: boolean;
  supplier_name?: string | null;
}

export interface RecipeListItem {
  id: number;
  name: string;
  selling_price: number;
  category: string | null;
  target_margin: number | null;
  food_cost: number | null;
  food_cost_percent: number | null;
  is_homemade: boolean;
  margin_status: MarginStatus;
  created_at: string;
}

export interface RecipeDetail extends RecipeListItem {
  ingredients: RecipeIngredientResponse[];
}

export interface RecipeListResponse {
  items: RecipeListItem[];
  total: number;
}

export interface DashboardResponse {
  average_food_cost_percent: number | null;
  total_recipes: number;
  recipes_green: number;
  recipes_orange: number;
  recipes_red: number;
  recipes: RecipeListItem[];
}

export interface RecipeIngredientInput {
  ingredient_id: number;
  quantity: number;
  unit: string;
}

export interface RecipeCreateRequest {
  name: string;
  selling_price: number;
  category?: string | null;
  target_margin?: number | null;
  is_homemade?: boolean;
  ingredients: RecipeIngredientInput[];
}

export interface RecipeUpdateRequest {
  name?: string;
  selling_price?: number;
  category?: string | null;
  target_margin?: number | null;
  is_homemade?: boolean;
  ingredients?: RecipeIngredientInput[];
}
